larray/core/constrained.py

```diff
@@ -0,0 +1,533 @@
+
+            # --- undeclared variables
+            for name, value in input_data.items():
+                self.__setattr__(name, value, skip_allow_mutation=True)
+
+        # code of the method below has been partly borrowed from pydantic.BaseModel.__setattr__()
+        def _check_key_value(self, name: str, value: Any, skip_allow_mutation: bool, skip_validation: bool) -> Any:
+            config = self.__config__
+            if not config.extra and name not in self.__fields__:
+                raise ValueError(f"Variable '{name}' is not declared in '{self.__class__.__name__}'. "
+                                 f"Adding undeclared variables is forbidden. "
+                                 f"List of declared variables is: {list(self.__fields__.keys())}.")
+            if not skip_allow_mutation and not config.allow_mutation:
+                raise TypeError(f"Cannot change the value of the variable '{name}' since '{self.__class__.__name__}' "
+                                f"is immutable and does not support item assignment")
+            known_field = self.__fields__.get(name, None)
+            if known_field:
+                if not skip_validation:
+                    value, error_ = known_field.validate(value, self.dict(exclude={name}), loc=name, cls=self.__class__)
+                    if error_:
+                        raise error_.exc
+            else:
+                warnings.warn(f"'{name}' is not declared in '{self.__class__.__name__}'", stacklevel=3)
+            return value
+
+        def __setitem__(self, key, value, skip_allow_mutation=False, skip_validation=False):
+            if key != 'meta':
+                value = self._check_key_value(key, value, skip_allow_mutation, skip_validation)
+                # we need to keep the attribute in sync
+                object.__setattr__(self, key, value)
+                self._objects[key] = value
+
+        def __setattr__(self, key, value, skip_allow_mutation=False, skip_validation=False):
+            if key != 'meta':
+                value = self._check_key_value(key, value, skip_allow_mutation, skip_validation)
+                # we need to keep the attribute in sync
+                object.__setattr__(self, key, value)
+            Session.__setattr__(self, key, value)
+
+        def __getstate__(self) -> Dict[str, Any]:
+            return {'__dict__': self.__dict__}
+
+        def __setstate__(self, state: Dict[str, Any]) -> None:
+            object.__setattr__(self, '__dict__', state['__dict__'])
+
+        def dict(self, exclude: Set[str] = None):
+            d = dict(self.items())
+            for name in exclude:
+                if name in d:
+                    del d[name]
+            return d
+
+    class Parameters(ConstrainedSession):
+        """
+        Same as py:class:`ConstrainedSession` but:
+
+            * declared variables cannot be modified after initialization
+            * adding undeclared variables after initialization is forbidden.
```

Comment:
Unsure this second point should be allowed in ConstrainedSession either. What does your experience with playing with it gives in this regard?

Reply:
My idea was to "force" our users to declare all parameters in one place (discipline! discipline! discipline!).
But on the other hand, our users don't like to be forced (even encouraged) to follow good practices.
Not sure if it is actually a true good practice (I learned it was but is it true?).

Now after a second thought, I guess the best thing to do considering our users is to be more permissive in the implementation of LArray but showing the best way to use LArray objects in the tutorial.